Styled components 4

dr3 opened the pull request:
Resolves #1040 

Upgrades to Styled Components v4.1.2

- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

Commit:
Upgrade to styled-components 4

Commit:
Update majority of snapshots

Commit:
Update global styles method

Commit:
Update style tests for the new SC version

Commit:
Update GlobalStyles test

Commit:
Merge branch 'latest' into StyledComponents4

ChrisBAshton reviewed (dismissed):
LGTM 👍 agreed `ForwardRef` doesn't look great, but is an unavoidable thing at the moment and we'll extract out in the move to Enzyme or react-test-library soon...

sareh commented:
✅ Tested the storybook implementation
- `npm ci && npm run storybook` 
- visited http://localhost:9001/,
Inspected the components and saw that the global styles were being applied correctly

<img width="1355" alt="screen shot 2018-12-04 at 15 24 59" src="https://user-images.githubusercontent.com/3028997/49452320-e6ae9d80-f7d8-11e8-94f1-712272c80939.png">

:x: Tested the application implementation
- `npm run dev` 
- visited http://localhost:7080/news/articles/c9rpqy7pmypo
The global styles seem to be added after the majority of the component styles:

<img width="1131" alt="screen shot 2018-12-04 at 15 32 33" src="https://user-images.githubusercontent.com/3028997/49452794-c03d3200-f7d9-11e8-86c2-12c6a7b68ab4.png">
Note the comment: `sc-global` indicating where the global styles start.


:x: Testing the application implementation in the prod version:
- `npm run build && npm run start`
- visited http://localhost:7080/news/articles/c9rpqy7pmypo
- I needed to turn JavaScript off, to see the styles in the `<style>` tag in the head: 
Same issue with the global styles being imported near the end of the list of items. 

I think it's important to investigate why the Global Styles are imported in the order that they are.  This is likely to do with how we use Helmet in our Document component.

sareh commented on src/app/components/Document/index.jsx:
```diff
@@ -32,6 +32,7 @@ const Document = ({ assets, app, data, styleTags, helmet }) => {
         {meta}
         <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon" />
         <link rel="manifest" href="/manifest.json" />
+        <GlobalStyle />
```
I'd suggest putting the global styles below the resource hints. This is so that the browser can parse the resource hints before getting to any of the styles.

Commit:
Move global styles below resource hints

Commit:
Merge branch 'latest' of github.com:BBC-News/simorgh into StyledComponents4

Commit:
Update snapshot

Commit:
Merge branch 'StyledComponents4' of github.com:BBC-News/simorgh into StyledComponents4

sareh commented:
The point I raised in this comment https://github.com/BBC-News/simorgh/pull/1042#issuecomment-444144485 about the Global Styles being rendered after other components' styles, rather than being the first in the style tag, is solely due to how Helmet handles `styleTags` that are passed into the application. 

Even though `<GlobalStyle>` is imported in the `Document` on this line https://github.com/BBC-News/simorgh/pull/1042/files#diff-00a291acea0cc37f44c9f1daaa1fe90cR36 that is not where the styles are imported.  

I'll open another issue that focuses on addressing how to ensure global styles are in a style tag prior to the component styles, which is very relevant to the cascade and how CSS works. 

In the meantime, for clarity for developers looking at this code and refactoring it, it would be better to have the `<GlobalStyle>` at the top of the `Articles` container's render function.  https://github.com/BBC-News/simorgh/blob/latest/src/app/containers/Article/index.jsx#L74

Commit:
Move GlobalStyles to top of article

Commit:
Pull Latest

sareh reviewed (dismissed):
👍 Looks good!

ChrisBAshton committed:
Merge branch 'latest' into StyledComponents4

ChrisBAshton committed:
removes array casting, fixes #976

Upgrade to styled-components v4 means we don't need to cast our getStyleElement() call to an array anymore. It consistently returns an array.

Commit:
Merge pull request #1046 from BBC-News/simplify-amp-custom-logic

removes array casting, fixes #976

sareh reviewed (approved):
Tested locally and it looks good. Global Styles all in the right place.

jamesbrumpton commented:
LGTM
